homeassistant/components/modbus/bit_sensor.py

```diff
@@ -0,0 +1,223 @@
+from __future__ import annotations
+
+from datetime import timedelta
+from functools import lru_cache, partial
+import logging
+import time
+
+from pymodbus.exceptions import ConnectionException, ModbusException
+from pymodbus.pdu import ExceptionResponse
+
+from homeassistant.components.binary_sensor import BinarySensorEntity
+from homeassistant.const import (
+    CONF_ADDRESS,
+    CONF_COUNT,
+    CONF_DEVICE_CLASS,
+    CONF_NAME,
+    CONF_SCAN_INTERVAL,
+    CONF_SLAVE,
+    STATE_ON,
+)
+from homeassistant.helpers.event import async_track_time_interval
+from homeassistant.helpers.restore_state import RestoreEntity
+from homeassistant.helpers.typing import DiscoveryInfoType, HomeAssistantType
+
+from .const import (
+    CALL_TYPE_REGISTER_INPUT,
+    CONF_BIT_NUMBER,
+    CONF_BIT_SENSORS,
+    CONF_INPUT_TYPE,
+    MODBUS_DOMAIN,
+)
+from .modbus import ModbusHub
+
+_LOGGER = logging.getLogger(__name__)
+
+
+@lru_cache(maxsize=32)
+def _read_cached(hub, method, ttl_bucket, *args, **kvargs):
+    """Return cached or invoke the Hub read_* method."""
+    return getattr(hub, method)(*args, **kvargs)
+
+
+class ModbusReadCache:
+    """Wraps Modbus Hub and provide cached methods."""
+
+    CACHED_METHODS = ["read_input_registers", "read_holding_registers"]
+    CACHE_RESET_METHODS = "write_"
+
+    def __init__(self, hub):
+        """Init the read cache."""
+        self._hub = hub
+
+    def __getattr__(self, attr):
+        """Forward calls to the Hub object or use cached."""
+        if attr.startswith(ModbusReadCache.CACHE_RESET_METHODS):
+            _read_cached.cache_clear()
+        if attr not in ModbusReadCache.CACHED_METHODS:
+            return getattr(self._hub, attr)
+
+        return partial(_read_cached, self._hub, attr, int(time.time()))
+
+
+def setup_bit_sensors(
+    hass: HomeAssistantType,
+    discovery_info: DiscoveryInfoType | None = None,
+) -> [BinarySensorEntity]:
+    """Set up the Modbus Bit sensors."""
+    sensors = []
+
+    if discovery_info is None:
+        return sensors
+
+    for entry in discovery_info.get(CONF_BIT_SENSORS, []):
+        words_count = int(entry[CONF_COUNT])
+        bit_number = int(entry[CONF_BIT_NUMBER])
+
+        if bit_number >= words_count * 16:
+            _LOGGER.error(
+                "Bit number for the %s sensor is out of range",
+                entry[CONF_NAME],
+            )
+            continue
+
+        hub: ModbusHub = hass.data[MODBUS_DOMAIN][discovery_info[CONF_NAME]]
+        sensors.append(
+            ModbusBitSensor(
+                hub,
+                entry[CONF_NAME],
+                entry.get(CONF_SLAVE),
+                entry[CONF_ADDRESS],
+                entry[CONF_INPUT_TYPE],
+                bit_number,
+                words_count,
+                entry.get(CONF_DEVICE_CLASS),
+                entry[CONF_SCAN_INTERVAL],
+            )
+        )
+
+    return sensors
+
+
+class ModbusBitSensorBase(RestoreEntity, BinarySensorEntity):
```

Comment:
Why do you need a base ? Seems like you could merge ModbusBitSensor and MobusBitSensorBase.